forms/base_form.py

```diff
@@ -0,0 +1,50 @@
+from flask_wtf import FlaskForm
+from wtforms.validators import ValidationError
+
+from lib.util import humanize_camelcased_string
+
+
+class BaseForm(FlaskForm):
+    def get_template(self, subclass_name):
+        return getattr(self.__class__, subclass_name)()
+
+    @property
+    def error_messages(self):
+        return list(_iterate_error_messages(prefixes=[], errors=self.errors))
+
+    def _validate_uniqueness(self, message, value_list):
+        seen       = set()
+        duplicated = set()
+
+        for value in value_list:
+            if value in seen:
+                duplicated.add(value)
+            else:
+                seen.add(value)
+
+        if duplicated:
+            value_description = ', '.join([repr(d) for d in duplicated])
+            raise ValidationError(f"{message}: {value_description}")
+
+
+def _iterate_error_messages(prefixes, errors):
```

Comment:
This is a fairly complicated piece of code, to understand it, it's recommended to look up the `BaseForm` class' test where there's examples. The goal is to extract readable error messages from the nested forms coming from [WTForms](https://wtforms.readthedocs.io/en/3.2.x/).